### ISSUE TITLE
PropertyPlaceholder not resolved in application.yml

### ISSUE THREAD

**Issue by frehov** [micronaut-projects/micronaut-core#839]
Micronaut is not resolving nested properties, even when the property is defined before/after offending property that is not resolving.

### Task List

- [X] Steps to reproduce provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Clone repo https://github.com/frehov/micronaut-swagger-server
2. Change property `application.api.swagger,urls[0].url` to `${application.api.swagger.path}/swagger-server-0.1.yml`
3. Run application and go to localhost:9090/api

### Expected Behaviour
property defined in another property should be resolved before injecting/binding.  
In this case the property should resolve to `/api-docs/swagger/swagger-server-0.1.yml`

### Actual Behaviour

Nested placeholder is not resolved before binding/injection
![image](https://user-images.githubusercontent.com/37329363/47947373-9dd6a100-df1b-11e8-957d-773b4b80d4bf.png)

### Environment Information

- **Operating System**: Windows 10
- **Micronaut Version:** 1.0.0
- **JDK Version:** 1.8.0_161

### Example Application

-  https://github.com/frehov/micronaut-swagger-server

**Comment by frehov** [comment]
After debugging my way through the code responsible for resolving proerties, and looking at what happens, It's obvious that nested properties for maps in lists are not processed.

The code where resolving fails: https://github.com/micronaut-projects/micronaut-core/blob/01e2b1ef1255b584e9e5c1965d334e1190e2ade5/inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java#L407

To resolve nested properties contained within lists, I believe you'll need to check for instance of List, and then resolve each property inside, taking care to check whether the entry inside is a List, Map or a String.

The propertyResolver does not resolve the follwing either, even though the property is present in the propertymap
![image](https://user-images.githubusercontent.com/37329363/47951988-cdf86100-df68-11e8-900a-f7d3f88e3e40.png)


so to summarize:
1. Check for nested properties in the value returned from a property
2. Check for nested properties inside Lists, as these can subsequently be either List, Map or String

**Comment by frehov** [comment]
the following test will break the build if added to PropertySourcePropertyResolverSpec

```
void "test resolve placeholders for lists of map"() {
        given:
        def values = [
                'foo.bar' : '10',
                'foo.bar.list' : [
                        [
                                'foo' : '${foo.bar}'
                        ],
                        [
                                'bar' : 'baz'
                        ]
                ]
        ]

        PropertySourcePropertyResolver resolver = new PropertySourcePropertyResolver(
                PropertySource.of("test", values)
        )

        expect:
        resolver.getProperty('foo.bar.list', List).isPresent()
        resolver.getProperty('foo.bar.list', List).get() == [['foo': '10'], ['bar': 'baz']]
}
```

**Comment by frehov** [comment]
Just pulled your changes and the test does indeed pass now.

However, another legal yml config is lists inside lists ... etc.

```
void "test resolve placeholders for lists of lists with maps"() {
        given:
        def values = [
                'foo.bar' : '10',
                'foo.bar.list' : [[
                        [
                                'foo' : '${foo.bar}'
                        ],
                        [
                                'bar' : 'baz'
                        ]
                ]]
        ]

        PropertySourcePropertyResolver resolver = new PropertySourcePropertyResolver(
                PropertySource.of("test", values)
        )

        expect:
        resolver.getProperty('foo.bar.list', List).isPresent()
        resolver.getProperty('foo.bar.list', List).get() == [['foo': '10'], ['bar': 'baz']]
    }
```

I can open a separate issue on this if you'd like?

**Comment by cchandurkar** [comment]
@frehov do you know if this issue is resolved? I'm on 3.9.4 and still seeing that nested configuration properties are not being able to resolve environment variables properly.
